### PR TITLE
feat: apply patches through translate_package_lock

### PR DESCRIPTION
**Pull request by kormide**

### CODE REVIEW

File: WORKSPACE [hunk 1]
```diff
@@ -53,6 +53,16 @@ npm_import(
...
+    patch_args = {
+        "@gregmagolan/test-a@0.0.1": [
+            "-p1",
+        ],
+    },
+    patches = {
+        "@gregmagolan/test-a@0.0.1": [
+            "//example:test-a.patch",
+        ],
+    },
```
Comment (kormide) [inline WORKSPACE:65]: Planning to also support wildcards `"*"` to patch everything (ngcc case), and the package name minus the version to patch all version. But will do that in a separate PR.

File: js/npm_import.bzl [hunk 2]
```diff
@@ -1,6 +1,7 @@
 "repository rules for importing packages from npm"
 
 load("//js/private:translate_package_lock.bzl", lib = "translate_package_lock")
+load("@bazel_tools//tools/build_defs/repo:utils.bzl", "patch")
```
Comment (kormide) [inline js/npm_import.bzl:4]: This is what `http_archive` uses and unlike the `repository_ctx` is supports `patch_args` for git patches as well as other stuff.

File: js/private/translate_package_lock.bzl [hunk 3]
```diff
@@ -105,7 +116,7 @@ _NPM_IMPORT_TMPL = """\
         integrity = "{integrity}",
         package = "{package}",
         version = "{version}",
-        deps = {deps},
+        deps = {deps},{maybe_patches}{maybe_patch_args}
```
Comment (kormide) [inline js/private/translate_package_lock.bzl:119]: A bit awkward, but I didn't want to output `patches = [...]` and `patch_args = [...]` if there weren't any.

**Review by alexeagle — APPROVED** [review]

File: js/private/translate_package_lock.bzl (hunk 3 above)
Comment (alexeagle) [reply js/private/translate_package_lock.bzl:119]: yeah I was commenting about it but came to the same conclusion, sometimes the generated thing will be read

File: WORKSPACE (hunk 1 above)
Comment (kormide) [reply WORKSPACE:65]: Actually I think the wildcard is only applicable for postinstall, not patches.

**Comment by kormide** [comment]
I added a second commit to support patching all versions of a particular package.

**Review by gregmagolan — APPROVED** [review]
Beauty

File: js/private/translate_package_lock.bzl [hunk 4]
```diff
@@ -140,13 +155,25 @@ def npm_repositories():
         if "requires" in dep.keys():
             for n in dep["requires"].keys():
                 deps.append("@" + _repo_name(n, [d["version"] for (p, d) in lockfile["dependencies"].items() if p == n][0]))
+
+        name_with_version = _repo_name_and_version(name, dep["version"])
+
+        patches = patches_map.get(name, [])[:]
+        patches.extend(patches_map.get(name_with_version, []))
+
+        patch_args = patch_args_map.get(name, [])
```
Comment (gregmagolan) [inline js/private/translate_package_lock.bzl:164]: nit: I would make this symmetrical here with
```
patch_args.extend(patch_args.get(name_with_version, []))
```
Not likely to be used but it keeps the semantics for the two dicts the same

Comment (kormide) [reply js/private/translate_package_lock.bzl]: Updated

**Review by gregmagolan — APPROVED** [review]
:taco:
